boltdesignsystem/bolt
# Fix block() and include() autoescaping

## Jira

http://vjira2:8080/browse/BDS-1988

## Summary

Fix twig `block()` and `include()` function usage to work with autoescaping

## Details
This PR has two parts:
1. Replace the twig `include()` function in the blueprints package with a macro that prints the contents of the include in order to play nice with autoescaping (credit goes to @sghoweri, thanks!)
2. Update places where the twig `block()` syntax is used to ensure that the block is always printed before setting its contents to a variable (also to play nice with autoescaping and remove the need for the `|raw` filter).  This can take several forms:
    - Most simply, ensuring that where the block contents is printed, the ` {{ block('block-name') }}` syntax (and not a printed variable) is used.
    - In more complex cases-- especially when embeds result in multiple layers of blocks, potentially with conflicting names-- setting a variable representing the block and its contents using the following syntax:
```
{% if block("body") %}
  {% set body_block %}
    {{ block("body") }}
  {% endset %}
{% endif %}
```

## How to test
- Confirm that replacement card, toolbar, and all blueprint patterns can be printed with autoescaping enabled (in particular, in Drupal) without raw markup displaying.  Unfortunately, Drupal lab needs some updates to be functional, which I omitted from this PR for clarity (but will work on in separate PRs).
- Confirm no regressions in pattern lab.  In particular, confirm that when a block is either not present or empty, any conditional wrappers do not print.  For example, both the following should result in cards that do not have a `<bolt-card-replacement-media>` element within them

```
{% embed "@bolt-components-card-replacement/card-replacement.twig"  only %}
  {% block body %}
    hello world
  {% endblock %}
{% endembed %}
```
```
{% embed "@bolt-components-card-replacement/card-replacement.twig" only %}
  {% block media %}{% endblock %}

  {% block body %}
    hello world
  {% endblock %}
{% endembed %}
```
